HPXMLtoOpenStudio/resources/airflow.rb

```diff
@@ -1636,41 +1561,105 @@ def self.apply_infiltration_and_ventilation_fans(model, weather, vent_mech, vent
+        infil_program.addLine('Set Qexhaust = Qexhaust+QWHV')
+      elsif [HPXML::MechVentTypeSupply, HPXML::MechVentTypeCFIS].include? vent_mech.fan_type
+        infil_program.addLine('Set Qsupply = Qsupply+QWHV')
+      elsif [HPXML::MechVentTypeBalanced, HPXML::MechVentTypeERV, HPXML::MechVentTypeHRV].include? vent_mech.fan_type
+        infil_program.addLine('Set Qexhaust = Qexhaust+QWHV')
+        infil_program.addLine('Set Qsupply = Qsupply+QWHV')
+      end
     end
 
-    infil_program.addLine('Set Q_tot_flow = (((Qu^2)+(Qn^2))^0.5)')
-    infil_program.addLine('Set Q_tot_flow = (@Max Q_tot_flow 0)')
+    # Calculate adjusted infiltration based on mechanical ventilation system
+    infil_program.addLine('Set Qfan = (@Max Qexhaust Qsupply)')
+    if Constants.ERIVersions.index(@eri_version) >= Constants.ERIVersions.index('2019')
+      # Follow ASHRAE 62.2-2016, Normative Appendix C equations for time-varying total airflow
+      infil_program.addLine('If Qfan > 0')
+      # Balanced system if the total supply airflow and total exhaust airflow are within 10% of their average.
+      infil_program.addLine('  Set Qavg = ((Qexhaust + Qsupply) / 2.0)')
+      infil_program.addLine('  If ((@Abs (Qexhaust - Qavg)) / Qavg) <= 0.1') # Only need to check Qexhaust, Qsupply will give same result
+      infil_program.addLine('    Set phi = 1')
+      infil_program.addLine('  Else')
+      infil_program.addLine('    Set phi = (Qinf / (Qinf + Qfan))')
+      infil_program.addLine('  EndIf')
+      infil_program.addLine('  Set Qinf_adj = phi * Qinf')
+      infil_program.addLine('Else')
+      infil_program.addLine('  Set Qinf_adj = Qinf')
+      infil_program.addLine('EndIf')
+    else
+      infil_program.addLine('Set Qu = (@Abs (Qexhaust - Qsupply))') # Unbalanced flow
+      infil_program.addLine('Set Qb = Qfan - Qu') # Balanced flow
+      infil_program.addLine('Set Qtot = (((Qu^2) + (Qinf^2)) ^ 0.5) + Qb')
+      infil_program.addLine('Set Qinf_adj = Qtot - Qu - Qb')
+    end
 
     # Assign total airflow to different component loads
-    infil_program.addLine("Set #{imbal_mechvent_flow_actuator.name} = (@Abs ((Qout - QductsOut) - (Qin - QductsIn)))")
-    infil_program.addLine("Set #{imbal_ducts_flow_actuator.name} = (@Abs (QductsOut - QductsIn))")
-
-    # Assign the remainder to infiltration:
-    infil_program.addLine("Set #{infil_flow_actuator.name} = Q_tot_flow - #{imbal_mechvent_flow_actuator.name} - #{imbal_ducts_flow_actuator.name}")
-    infil_program.addLine("If #{infil_flow_actuator.name} < 0")
-    infil_program.addLine("  Set #{infil_flow_actuator.name} = 0")
-    infil_program.addLine('EndIf')
+    if (not vent_mech.nil?) && [HPXML::MechVentTypeERV, HPXML::MechVentTypeHRV].include?(vent_mech.fan_type)
+      infil_program.addLine("Set #{mechvent_flow_actuator.name} = Qfan - QWHV") # QWHV load captured by ERV/HRV program below
+    else
+      infil_program.addLine("Set #{mechvent_flow_actuator.name} = Qfan")
```

Comment:
Yes, you mostly have it right. Though when you say "both balanced and unbalanced is addressed" in the new approach, I think that's not 100% accurate. It's slightly more nuanced...

Example A: Suppose for an hour that you have supply fan = 50 cfm and exhaust kitchen fan = 55 cfm. With ASHRAE 62.2, it's treated as 55 cfm **balanced** because they are within 10% of the average. Seems reasonable enough.

Example B: Now suppose for a given hour that you have supply fan = 50 cfm and exhaust kitchen fan = 65 cfm. With ASHRAE 62.2, it's treated as 65 cfm **imbalanced** (as opposed to 50 cfm balanced, 15 cfm imbalanced). 

So ASHRAE 62.2 is binary on a timestep basis: it's either treated as fully balanced or fully imbalanced, where it is treated as fully imbalanced up to the point where supply and exhaust airflows are within 10% of their average.

So yes, I think this is definitely better than before, and I'm glad you agree! But it still seems a little imprecise to me. Practically speaking however, it's unlikely that you will have many hours with significant balanced airflow, so I think it's a reasonable approximation.